#8233 Restrict API key usage by source IP

PieterL75 opened the pull request:
<!--
    Thank you for your interest in contributing to NetBox! Please note
    that our contribution policy requires that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
    IF YOUR PULL REQUEST DOES NOT REFERENCE AN ACCEPTED BUG REPORT OR
    FEATURE REQUEST, IT WILL BE MARKED AS INVALID AND CLOSED.
-->
### Fixes: #8233 Restrict API key usage by source IP
<!--
    This PR will add the capability of limiting the Source IP's that can use an API key.
-->

Commit:
base_requirements - django_better_admin_arrayfield

Commit:
#8233 Restrict API key usage by source IP

Commit:
Add 'Any' display to admin form

jeremystretch commented on netbox/users/models.py:
```diff
@@ -227,6 +238,24 @@ def is_expired(self):
+        except ValueError:
+            raise ValidationError(f"{raw_ip_address} is an invalid IP address")
+
+        for ipnet in self.allowed_ips:
+            if ip_address in ipaddress.ip_network(ipnet):
```
Try to avoid using overly terse variable names; they can impede comprehension.
```suggestion
        for ip_network in self.allowed_ips:
            if ip_address in ipaddress.ip_network(ip_network):
```

PieterL75 commented on netbox/netbox/api/authentication.py:
```diff
@@ -18,6 +23,20 @@ def authenticate_credentials(self, key):
 
+        # Verify source IP is allowed
+        request = self.request
+        if token.allowed_ips and request:
```
Yes. I do super of the authenicate(), that super calls the authenticate_credentials without a request parameter.
After the super I call the authenticate_credentials with 'request'...

hm.. maybe I need to move the source-ip check in the authenticate()... I needed the token, but I get that from the super...

Going to test that out..

PieterL75 committed:
rename ipnet var to ip_networks

Co-authored-by: Jeremy Stretch <jstretch@ns1.com>

PieterL75 commented on base_requirements.txt:
```diff
@@ -113,3 +113,7 @@ svgwrite
+
+# It changes comma separated widget to list based in admin panel
+# https://github.com/gradam/django-better-admin-arrayfield
+django_better_admin_arrayfield
```
The addon provides an easier way to add objects to a comma separated list.. The only drawback is, that this ony is for the Admin side... not for the Userside.. 
I'll remove it for now, but I think this kind of widget is a good addition to both admin and user gui

jeremystretch replied:
I understand; however every dependency that we introduce brings with it additional burdens. And smaller packages like these are highly susceptible to abandonment, which creates more work for us in the long-term.

PieterL75 commented on netbox/netbox/api/authentication.py:
```diff
@@ -18,6 +23,20 @@ def authenticate_credentials(self, key):
+            elif 'REMOTE_ADDR' in request.META:
+                clientip = request.META['REMOTE_ADDR']
+            else:
+                raise exceptions.AuthenticationFailed(f"The request HTTP headers (HTTP_X_REAL_IP, REMOTE_ADDR) are missing or do not contain a valid source IP.")
```
I actually don't test if the request has a valid ip. Only if the httpheaders exist..
Removed the " or do not contain a valid source IP" part of the message

Commit:
Remove betteradminfield
Move IP verification to authenticate()

PieterL75 committed:
Merge branch 'develop' into develop

Commit:
Remove betteradminarrat from requirements.txt

Commit:
Merge branch 'develop' of https://github.com/PieterL75/netbox into develop

Commit:
Fix authentication() returning None value

jeremystretch commented on netbox/netbox/api/authentication.py:
```diff
@@ -11,6 +11,28 @@ class TokenAuthentication(authentication.TokenAuthentication):
+                if not token.validate_client_ip(clientip):
+                    raise exceptions.AuthenticationFailed(f"Source IP {clientip} is not allowed to use this token.")
+
+            return token_user, token
+        else:
+            return None
```
```suggestion
    return authenticationresult
```

jeremystretch commented on netbox/users/models.py:
```diff
@@ -227,6 +237,24 @@ def is_expired(self):
+        try:
+            ip_address = ipaddress.ip_address(raw_ip_address)
+        except ValueError:
+            raise ValidationError(f"{raw_ip_address} is an invalid IP address")
```
I feel like we should return a 400 Bad Request response if the IP address is invalid. (We can handle this under `authenticate()` by catching ValidationError. Otherwise, raising `ValidationError` (which AFAICT won't be caught by DRF) is just going to trigger a 500 so we should figure out something a bit cleaner.

PieterL75 replied:
Not sure how to do this.. Any example code that accomplish this ?

jeremystretch replied:
We should be able to catch the `ValidationError` and return a [`HttpResponseBadRequest`](https://docs.djangoproject.com/en/3.2/ref/request-response/#django.http.HttpResponseBadRequest) response, with an error message indicating the invalid value.

PieterL75 replied:
Hm.. not easy :-)
Do I need to extend the ExceptionHandlingMiddleware to catch the error and generate a 400 ?
a simple `return HttpResponseBadRequest(f"ValidationErrorInfo: {ValidationErrorInfo.message}")` just throws another error that the return of authenticate() expects 2 values, but does not return the 404 error to the client

PieterL75 replied:
Found it :
```
except ValidationError as ValidationErrorInfo:
    raise exceptions.ValidationError(f"The Source IP in the HTTP Header {http_header} has a ValidationError: {ValidationErrorInfo.message}")
```

This raises a `HTTP/1.1 400 Bad Request`

PieterL75 replied:
@jeremystretch Does this solution seems good ?

jeremystretch commented on netbox/users/models.py:
```diff
@@ -227,6 +237,24 @@ def is_expired(self):
 
+    def validate_client_ip(self, raw_ip_address):
+        """
+        Checks that an ip address falls within the allowed ips.
```
```suggestion
        Checks that an IP address falls within the allowed IPs.
```

jeremystretch commented on netbox/users/admin/__init__.py:
```diff
@@ -58,9 +58,15 @@ def get_inlines(self, request, obj):
     ]
 
+    def list_allowed_ips(self, obj):
+        if obj.allowed_ips:
+            return obj.allowed_ips
+        return 'Any'
```
```suggestion
        return obj.allowed_ips or 'Any'
```

PieterL75 replied:
Nice ! Whish I knew how to code this clean :-)

PieterL75 committed:
Beautify comments

Co-authored-by: Jeremy Stretch <jstretch@ns1.com>

PieterL75 committed:
Optimize returning allowed_ips or 'Any'

Co-authored-by: Jeremy Stretch <jstretch@ns1.com>

PieterL75 committed:
Optimize return values

Co-authored-by: Jeremy Stretch <jstretch@ns1.com>

Commit:
Correct "'return' outside function"

Commit:
Raise 400 validation error on bad clientip

PieterL75 committed:
Merge branch 'develop' into develop

Commit:
Save old JobResults

Commit:
Add dynamic config JOBRESULT_RETENTION
and cleanup functionality to the housekeeping script

jeremystretch committed:
Merge branch 'feature' into develop

jeremystretch commented:
Should be all set now.

PieterL75 commented:
@jeremystretch its failing on the documentation build.. anything you can fix plz ?

Commit:
Limit amount of nonracked devices displayed

Fixes #8920

Limits the amount of non-racked devices on Site and Location view to 10 and provides a link to the device list this is pre-filtered to the relevant site or location.

Commit:
Update nonracked_devices.html

Fixing issue where Displaying 10 of 10 would appear when it's not required.

jeremystretch commented:
@PieterL75 sorry, I'm not sure what happened here. There are definitely still some discrepancies with the branch. I may have messed something up earlier.

Given that there's not a ton of change in this PR, it might be cleanest to create a new fork of the `feature` branch and just copy over the changes by hand from the diff here, then resubmit that as a new PR.

Additionally, the migration file will need to be bumped to `0003_token_allowed_ips` so as not to conflict with the v3.2 migration.

kkthxbye-code committed:
Change default JOBRESULT_RETENTION from 0 to 90

Commit:
Updating to use a single queryset

Updating to use a single queryset for both template variables

JoeIzzard committed:
Closes #8415: Adds Neutrik power connectors

jeremystretch committed:
Fixes #9100: Include position field in module type YAML export

jeremystretch committed:
Fixes #9096: Remove duplicate filter tag when filtering by "none"

jeremystretch committed:
#9096: Correct getattr() call

jeremystretch committed:
Tweak changelog for #9061

jeremystretch committed:
Closes #9099: Enable display of installed module serial & asset tag in module bays list

jeremystretch committed:
Closes #9081: Add fhrpgroup_id filter for IP addresses

jeremystretch committed:
Fixes #9116: assigned_to_interface filter for IP addresses should not match FHRP group assignments

jeremystretch committed:
Merge pull request #9067 from jasonyates/FR8920-nonrackeddevices

Limit amount of nonracked devices displayed

jeremystretch committed:
Merge pull request #8957 from kkthxbye-code/save-job-results

Fix #8956: Save old JobResults

jeremystretch committed:
Changelog for #8920, #8956

JoeIzzard committed:
fix: standardisation

jeremystretch committed:
Merge pull request #9110 from JoeIzzard/develop

Closes #8415: Adds Neutrik power connectors

jeremystretch committed:
Changelog for #9110

jeremystretch committed:
Fixes #9118: Fix validation error when importing VM child interfaces

Commit:
Add filters to VLAN selection group

jeremystretch committed:
Merge pull request #9124 from tyler-8/wlan_site_filter_3_2

Add filters to VLAN selection group for WirelessLAN Form

jeremystretch committed:
Changelog for #8543 & cleanup

jeremystretch committed:
Closes #9123: Improve appearance of SSO login providers

jeremystretch committed:
Fixes #9128: Resolve component labels per module bay position when installing modules

jeremystretch committed:
Release v3.2.1

jeremystretch committed:
Merge pull request #9131 from netbox-community/develop

Release v3.2.1

jeremystretch committed:
PRVB

jeremystretch committed:
Restore documentation search function

markkuleinio committed:
Require Python 3.8+ to run upgrade.sh

jeremystretch committed:
Merge pull request #9134 from markkuleinio/python-version

Fixes #9133: Require Python 3.8+ to run upgrade.sh

jeremystretch committed:
Changelog for #9133

jeremystretch committed:
Closes #9136: Add SSO configuration guide for Microsoft Azure AD

jeremystretch committed:
Add title for missing okta-openidconnect backend

jeremystretch committed:
Closes #9137: Add SSO configuration guide for Okta

jeremystretch committed:
Fixes #9156: Fix loading UserConfig data from fixtures

jeremystretch committed:
Fixes #9151: Child prefix counts not annotated on aggregates list under RIR view

jeremystretch committed:
Closes #9152: Annotate related object type under custom field view

jeremystretch committed:
Fixes #9132: Limit location options by selected site when creating a wireless link

jeremystretch committed:
Fixes #9158: Do not list tags field for CSV forms which do not support tag assignment

minitriga committed:
Closes #9060: Implement modulebay, iventory items and device bay filters (#9146)

* Closes #9060: Implement modulebay, iventory items and device bay filters

* add blank line

jeremystretch committed:
Changelog & test for #9060

PieterL75 committed:
Merge branch 'netbox-community:develop' into develop

PieterL75 commented:
As suggested by @jeremystretch : Closing this PR and recreating it with a new Fork of the feature branch
